IPython/core/interactiveshell.py

```diff
@@ -2711,12 +2713,13 @@ def get_cells():
         with prepended_to_syspath(dname):
             try:
                 for cell in get_cells():
-                    # self.run_cell currently captures all exceptions
-                    # raised in user code.  It would be nice if there were
-                    # versions of run_cell that did raise, so
-                    # we could catch the errors.
                     result = self.run_cell(cell, silent=True, shell_futures=shell_futures)
-                    if not result.success:
+                    if not result.success and raise_exceptions:
+                        if result.error_before_exec is not None:
+                            raise result.error_before_exec
+                        else:
+                            raise result.error_in_exec
+                    elif not result.success:
                         break
             except:
```

Comment:
This outer except will still catch any errors and display the traceback; for writing programmatic tests, the exception should escape the function call so the caller can catch it.